### DISCUSSION TITLE
where exactly is rofi_view_keyboard_navigation?

### DISCUSSION THREAD

**Discussion by ana-rchy** [davatorium/rofi#1635] [Q&A]
was trying to do add a keybind to rofi for an issue im working on, but came across this in a comment in view.c, but searching for it brings no results. and i suspect thats why my keybind isnt recognised in the first place after i had compiled my fork, but i have no clue where this is so yeah

**Comment by DaveDavenport** [comment]
I think that is a stale comment. you can ignore it.

**Reply by ana-rchy** [reply]
ill add its removal into my fork then, but id still like some guidance with actually getting the keybind recognized. would you be able to help me with that?

**Reply by DaveDavenport** [reply]
Not with the current information, 'it does not work' is not a lot go on .

**Reply by ana-rchy** [reply]
my changes can be seen here https://github.com/ana-rchy/rofi (dont mind the bad comments on the commits)

as far as i understand, keyb.c/.h define the data structure to hold all the keybindings, view.c handles the input and matching it to those keybinds, and filebrowser.c at around line 430 has a function that would go to the parent directory, but im not sure how it entirely works. currently i just wanna solve the problem of the keybinding not being recognized at all, as in it doesnt show up with `rofi -show keys`, and doesnt do anything in filebrowser mode. it all compiles just fine, but at runtime it effectively has 0 changes to rofi’s behaviour

**Reply by ana-rchy** [reply]
i should also add im not entirely sure how view.c connects eventually to filebrowser.c, just that it has something to do with `state`

**Reply by DaveDavenport** [reply]
your binding does not follow the naming scheme, that is why it does not show up (needs to start with kb-)

You made a mode specific binding but then make a global implementation to select the first row. This is not acceptable for a binding named 'parent directory'. Because it applies to all modes.

This also goes to the more fundamental problem, rofi and most of it interface is independent of the mode shown. 
(Modes can be added via plugins, etc.)  Having a single binding that only applies to one mode is something I want to avoid.


Maybe it is better to re-use the 'kb-accept-alt' binding for going one up?

**Reply by ana-rchy** [reply]
im not really sure how you would make a binding specifically for just one mode, im guessing something like [special_command](https://github.com/davatorium/rofi/blob/next/source/modes/filebrowser.c#L431) would be the answer but im not sure how its used or what it actually does. another idea is adding a function in filebrowser.c specifically for going up a parent directory with a keybinding, and passing the keybind pressed into it as a parameter (checking if its kb-accept-alt), but then that goes back to global implementations for independent modes, and im again not sure how `state` works either

**Reply by DaveDavenport** [reply]
```patch
diff --git a/source/modes/filebrowser.c b/source/modes/filebrowser.c
index 0e9b271c..aafdb07f 100644
--- a/source/modes/filebrowser.c
+++ b/source/modes/filebrowser.c
@@ -474,6 +474,16 @@ static ModeMode file_browser_mode_result(Mode *sw, int mretv, char **input,
     }
     retv = RELOAD_DIALOG;
   } else if ((mretv & MENU_CUSTOM_INPUT) && *input) {
+    if (special_command) {
+      GFile *new = g_file_get_parent(pd->current_dir);
+      if (new) {
+        g_object_unref(pd->current_dir);
+        pd->current_dir = new;
+        free_list(pd);
+        get_file_browser(sw);
+      }
+      return RESET_DIALOG;
+    }
     char *p = rofi_expand_path(*input);
     char *dir = g_filename_from_utf8(p, -1, NULL, NULL, NULL);
     g_free(p);

```

this implements going up under the ctrl-shift-enter keybinding (kb-accept-custom-alt)

**Reply by ana-rchy** [reply]
would you mind explaining why this works? kinda looks just like black magic to me cause i still dont know what `special_command`/`MENU_CUSTOM_INPUT` does or what it exactly represents other than just the idea of "input other than whats been specified"

i sorta understand whatevers _inside_ the if-statement so no need to explain that part, but would also be helpful if you wanna explain it (edit: also why the preceding else-if has `&& *input`, or when the function is given input/what its given, my best guess would be it passes the string name of the keybinding just used?)

**Reply by DaveDavenport** [reply]
aah the *input should probably be moved a bit. (it was just a quick and dirty patch).

```patch
diff --git a/source/modes/filebrowser.c b/source/modes/filebrowser.c
index 0e9b271c..6d5bcdfc 100644
--- a/source/modes/filebrowser.c
+++ b/source/modes/filebrowser.c
@@ -473,22 +473,34 @@ static ModeMode file_browser_mode_result(Mode *sw, int mretv, char **input,
       }
     }
     retv = RELOAD_DIALOG;
-  } else if ((mretv & MENU_CUSTOM_INPUT) && *input) {
-    char *p = rofi_expand_path(*input);
-    char *dir = g_filename_from_utf8(p, -1, NULL, NULL, NULL);
-    g_free(p);
-    if (g_file_test(dir, G_FILE_TEST_EXISTS)) {
-      if (g_file_test(dir, G_FILE_TEST_IS_DIR)) {
+  } else if ((mretv & MENU_CUSTOM_INPUT)) {
+    if (special_command) {
+      GFile *new = g_file_get_parent(pd->current_dir);
+      if (new) {
         g_object_unref(pd->current_dir);
-        pd->current_dir = g_file_new_for_path(dir);
-        g_free(dir);
+        pd->current_dir = new;
         free_list(pd);
         get_file_browser(sw);
-        return RESET_DIALOG;
       }
+      return RESET_DIALOG;
+    }
+    if (*input) {
+      char *p = rofi_expand_path(*input);
+      char *dir = g_filename_from_utf8(p, -1, NULL, NULL, NULL);
+      g_free(p);
+      if (g_file_test(dir, G_FILE_TEST_EXISTS)) {
+        if (g_file_test(dir, G_FILE_TEST_IS_DIR)) {
+          g_object_unref(pd->current_dir);
+          pd->current_dir = g_file_new_for_path(dir);
+          g_free(dir);
+          free_list(pd);
+          get_file_browser(sw);
+          return RESET_DIALOG;
+        }
+      }
+      g_free(dir);
+      retv = RELOAD_DIALOG;
     }
-    g_free(dir);
-    retv = RELOAD_DIALOG;
   } else if ((mretv & MENU_ENTRY_DELETE) == MENU_ENTRY_DELETE) {
     retv = RELOAD_DIALOG;
   }
```

special_command here maps to the `kb-*alt`  variants,    MENU_OK to the `kb-accept`  and MENU_CUSTOM_INPUT  to the `kb-accept-custom`.  The naming of the defines is inherited from very early versions that had limited uses.

**Reply by ana-rchy** [reply]
would the only way to assign more keybinds outside of those be to add more values to the `MenuReturn` enum?

**Reply by DaveDavenport** [reply]
yeah. it kinda is.

**Reply by ana-rchy** [reply]
actually, i wonder: could you instead set `state->retv` to be a bitwise xor of two preexisting `MenuReturn` values, and then check for that in `file_browser_mode_result` with smth like `(mretv & (MENU_X ^ MENU_Y)) == (MENU_X ^ MENU_Y)`? you could of course also just use a bitwise shift

i wanna try writing this out to see if it works so give me a min

**Reply by DaveDavenport** [reply]
I am not sure what you want to do, but I am not going to change this at this point in time as it would break plugins and bindings.

**Reply by ana-rchy** [reply]
just to clear up a few things: the reason for this: `state->retv = MENU_CUSTOM_INPUT | MENU_CUSTOM_ACTION;` on `ACCEPT_CUSTOM_ALT` is specifically to distinguish it from `ACCEPT_CUSTOM` by means of being able to bitwise-and with both `MENU_CUSTOM_ACTION` and `MENU_CUSTOM_INPUT`, which results in `special_command` being true and letting the whole parent directory keybind happen instead of just a regular file selection?

also, would i be right in saying `**input` is representing the string name of file? it seems to be implied by the line `char *p = rofi_expand_path(*input);`

**Reply by DaveDavenport** [reply]
First seems correct, it works this way for maintaining backwards compatibility of the mode interface.  At some point this will be redone, but it will mean breakage of bindings and plugins.


`*input` is a pointer that possibly points to string the user entered in the entry (input) box.

**Reply by ana-rchy** [reply]
to clarify what i was trying to do: i was gonna attempt to make a keybind for going to the previously visited directory, but i realized id need to add a variable to hold the path of the previous directory into `RofiViewState` and then pass and return that value to/from `file_browser_mode_result`

i wanna try to do this _myself_, as im trying to do this in other to actually gain experience in open-source contribution that isnt just fixing up typos
